File: Scala/src/main/scala/TextbookTrack/Chapter10/BA10D.scala
package TextbookTrack.Chapter10

object BA10D {
  case class Label(labels: Vector[Char]) {
    private val labelIndices: Map[Char, Int] = labels.iterator.zipWithIndex.toMap
    val size: Int = labels.length

    def apply(label: Char): Int = labelIndices(label)

    def zipWithIndex: Iterator[(Char, Int)] = labels.iterator.zipWithIndex
  }

  case class ProbabilityMatrix(rowLabels: Label, columnLabels: Label, probabilities: Vector[Vector[Double]]) {
    val nrRows: Int = rowLabels.size
    val nrCols: Int = columnLabels.size

    def apply(row: Char, col: Char): Double = probabilities(rowLabels(row))(columnLabels(col))
  }

  case class HMM(alphabet: Label, states: Label, transition: ProbabilityMatrix, emission: ProbabilityMatrix)

  private def collectLabels(line: String): Vector[Char] =
    line.toVector.filter(_.isLetter)

  private def readMatrix(reader: Iterator[String], k: Int): Vector[Vector[Double]] = {
    val _: String = reader.next()
    reader
      .take(k)
      .map(_.split("\\s").drop(1).map(_.toDouble).toVector)
      .toVector
  }

  private def readData(reader: Iterator[String]): (String, HMM) = {
    val string: String = reader.next()
    val _: String = reader.next()
    val alphabet = Label(collectLabels(reader.next()))
    val _: String = reader.next()
    val states = Label(collectLabels(reader.next()))
    val _: String = reader.next()
    val transition = ProbabilityMatrix(states, states, readMatrix(reader, states.size))
    val _: String = reader.next()
    val emission = ProbabilityMatrix(states, alphabet, readMatrix(reader, states.size))
    val hmm = HMM(alphabet, states, transition, emission)
    (string, hmm)
  }

  def calcProbabilityOfEmittedString(hmm: HMM, emittedString: String): Double = {
    val HMM(_, states, transition, emission) = hmm
    val nrCols: Int = emittedString.length
    val forwardMatrix: Array[Array[Double]] = Array.fill(states.size, nrCols)(0.0)
    val char: Char = emittedString(0)
    for {
      (state, ix) <- states.zipWithIndex
    } forwardMatrix(ix)(0) = emission(state, char) / states.size

    for {
      (char, jy) <- emittedString.drop(1).zipWithIndex
      (state, ix) <- states.zipWithIndex
    } {
      val emissionProb: Double = emission(state, char)
      val scores: Iterator[Double] =
        states
          .zipWithIndex
          .map{ case (previousState, k) => forwardMatrix(k)(jy) * transition(previousState, state) * emissionProb }
      forwardMatrix(ix)(jy + 1) = scores.sum
    }

    forwardMatrix.map(_.last).sum
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (emittedString, hmm): (String, HMM) = readData(reader)
    val result: Double = calcProbabilityOfEmittedString(hmm, emittedString)
    println(result)
  }
}

File: Python/textbook_track/chapter10/ba10d.py
# Compute the Probability of a String Emitted by an HMM
import sys

from collections import namedtuple

HMM = namedtuple('HMM', ['alphabet', 'states', 'transition', 'emission'])


class Label:
    def __init__(self, labels):
        self._labels = labels
        self._label_indices = {item: ix for ix, item in enumerate(labels)}
        self._size = len(labels)

    def __getitem__(self, label):
        return self._label_indices[label]

    def __repr__(self):
        return ', '.join(map(str, self._labels))

    def __iter__(self):
        yield from self._labels

    @property
    def labels(self):
        return self._labels

    @property
    def size(self):
        return self._size


class ProbabilityMatrix:
    def __init__(self, row_labels, column_labels, probabilities):
        self._row_labels = row_labels
        self._column_labels = column_labels
        self._probabilities = probabilities

    def __getitem__(self, label_pair):
        row_label, col_label = label_pair
        return self._probabilities[self._row_labels[row_label]][self._column_labels[col_label]]

    @property
    def nr_rows(self):
        return self._row_labels.size

    @property
    def nr_cols(self):
        return self._column_labels.size


def read_matrix(reader, k):
    _ = next(reader)
    matrix = []
    for _ in range(k):
        line = next(reader)
        probabilities = [float(item) for item in line.split()[1:]]
        matrix.append(probabilities)
    return matrix


def read_data(reader):
    string = next(reader).rstrip()
    _ = next(reader)
    alphabet = Label(next(reader).split())
    _ = next(reader)
    states = Label(next(reader).split())
    _ = next(reader)
    transition = ProbabilityMatrix(states, states, read_matrix(reader, states.size))
    _ = next(reader)
    emission = ProbabilityMatrix(states, alphabet, read_matrix(reader, states.size))
    hmm = HMM(alphabet=alphabet, states=states, transition=transition, emission=emission)
    return string, hmm


def calc_probability_of_emitted_string(hmm, emitted_string):
    _, states, transition, emission = hmm
    nr_cols = len(emitted_string)
    forward_matrix = [[0.0] * nr_cols for _ in range(states.size)]
    char = emitted_string[0]
    for ix, state in enumerate(states):
        forward_matrix[ix][0] = emission[state, char] / states.size

    for jy, char in enumerate(emitted_string[1:], start=1):
        for ix, state in enumerate(states):
            emission_prob = emission[state, char]
            scores = (forward_matrix[k][jy-1] * transition[previous_state, state] * emission_prob
                      for k, previous_state in enumerate(states))
            forward_matrix[ix][jy] = sum(scores)

    return sum(map(lambda x: x[-1], forward_matrix))


def main():
    reader = sys.stdin
    emitted_string, hmm = read_data(reader)
    result = calc_probability_of_emitted_string(hmm, emitted_string)
    print(result)


if __name__ == '__main__':
    main()
